### PR TITLE
feat: cached provider PoC

### PR DESCRIPTION
**Pull request by gakonst**
# Problem

We want to avoid repeating requests which we expect to have the same response. This is useful when used together with Forge's mainnet forking tests.

# Solution

Add a `Cache` which stores all request/responses pairs in-memory, which then proceeds to store them in disk when it gets Dropped

### CODE REVIEW

File: ethers-providers/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        let addr = Address::random();
+
+        assert!(provider.cache().unwrap().requests.is_empty());
+
+        mock.push(U256::from(100u64)).unwrap();
+        let res = provider.get_balance(addr, None).await.unwrap();
+        assert_eq!(res, 100.into());
+
+        assert!(!provider.cache().unwrap().requests.is_empty());
+        dbg!(&provider.cache());
```
Comment (mattsse) [inline ethers-providers/src/cache.rs:96]:
```suggestion
```

File: ethers-providers/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        let key = serde_json::to_string(&CachedRequest { method, params })?;
+        let value = serde_json::to_string(&response)?;
+        self.requests.insert(key, value);
+        Ok(())
+    }
+}
+
+impl Drop for Cache {
+    fn drop(&mut self) {
+        let file = match File::options().write(true).read(true).create(true).open(&self.path) {
```
Comment (mattsse) [inline ethers-providers/src/cache.rs:61]: this should be `File::create`? which

> Opens a file in write-only mode. reate a file if it does not exist, and will truncate it if it does.

File: ethers-providers/src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    path::PathBuf,
+};
+
+#[derive(Clone, Debug, Default)]
+/// Simple in-memory K-V cache using concurrent dashmap which flushes
+/// its state to disk on `Drop`.
+pub struct Cache {
+    path: PathBuf,
+    // serialized request / response pair
+    requests: DashMap<String, String>,
```
Comment (mattsse) [inline ethers-providers/src/cache.rs:16]: should this be <String, serde_json::Value>?

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by gakonst** [comment]
@brockelmore curious what you think on this approach?

**Comment by tynes** [comment]
Depending on how forge implements its caching strategy, it might be useful to have a way to clear the cache

**Comment by gakonst** [comment]
Closing - will make it part of the Forking provider on Foundry

**Pull request closed without merging**